app/Http/Controllers/Rpc/WalletController.php

```diff
@@ -24,72 +24,87 @@
 use Adshares\Adserver\Http\Controllers\Controller;
 use Adshares\Adserver\Utilities\AdsUtils;
 use Illuminate\Http\Request;
+use Illuminate\Http\Response;
 use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Log;
 use Illuminate\Support\Facades\Validator;
 
 class WalletController extends Controller
 {
     const FIELD_ADDRESS = 'address';
     const FIELD_AMOUNT = 'amount';
+    const FIELD_ERROR = 'error';
+    const FIELD_FEE = 'fee';
     const FIELD_MESSAGE = 'message';
     const FIELD_TO = 'to';
+    const FIELD_TOTAL = 'total';
     const FIELD_MEMO = 'memo';
     const VALIDATOR_RULE_REQUIRED = 'required';
 
     public function calculateWithdrawal(Request $request)
     {
+        $addressFrom = $this->getAdServerAdsAddress();
+        if (!AdsValidator::isAccountAddressValid($addressFrom)) {
+            Log::error("Invalid ADS address is set: ${addressFrom}");
+            return self::json([], Response::HTTP_INTERNAL_SERVER_ERROR);
+        }
+
         Validator::make($request->all(), [
             self::FIELD_AMOUNT => [self::VALIDATOR_RULE_REQUIRED, 'integer', 'min:1'],
             self::FIELD_TO => self::VALIDATOR_RULE_REQUIRED,
         ])->validate();
         $amount = $request->input(self::FIELD_AMOUNT);
         $addressTo = $request->input(self::FIELD_TO);
 
-        $addressFrom = $this->getAdserverAdsAddress();
-        if (!AdsValidator::isAccountAddressValid($addressFrom)
-            || !AdsValidator::isAccountAddressValid($addressTo)) {
+        if (!AdsValidator::isAccountAddressValid($addressTo)) {
             // invalid input for calculating fee
-            return self::json([], 422);
+            return self::json([self::FIELD_ERROR => 'invalid address'], Response::HTTP_UNPROCESSABLE_ENTITY);
         }
+
         $fee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
 
         $total = $amount + $fee;
         $resp = [
             self::FIELD_AMOUNT => $amount,
-            'fee' => $fee,
-            'total' => $total,
+            self::FIELD_FEE => $fee,
+            self::FIELD_TOTAL => $total,
         ];
 
         return self::json($resp);
     }
 
     public function withdraw(Request $request)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this function 4, down to the maximum allowed 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1142)